app/assets/javascripts/angular/services/annotation_refresher.js.erb

```diff
@@ -1,31 +1,73 @@
-angular.module("Prometheus.services").factory('AnnotationRefresher', ["$http", "VariableInterpolator", function($http, VariableInterpolator) {
-  return function(graph, scope) {
-    var tags = graph.tags.map(function(e) {
-      if (e.name) {
-        var n = VariableInterpolator(e.name, scope.vars);
-        return n.split(",").map(function(s) { return s.trim(); });
-      } else {
-        return "";
+angular.module("Prometheus.services").factory('AnnotationRefresher', ["$http", "$q", "$timeout", "VariableInterpolator", function($http, $q, $timeout, VariableInterpolator) {
+  var MINUTE = 1000*60;
+  var promises = {};
```

Comment:
What is the reason you want to clean up the cache. Reducing memory footprint / preventing memory leaks? In that case it would make more sense to ensure the the lookup dictionary can't contain more than x elements, by throwing out the oldest entry in case the dictionary is full. Currently it's still possible to add a lot of entries (plus timeout() functions) by changing the times a lot.

The cleanup timeout of 1 minute also feels quite arbitrary. Why not using the refresh interval instead?